src/hdmf/common/resources.py

```diff
@@ -466,63 +476,69 @@ def add_ref_term_set(self, **kwargs):
-                file = self._get_file_from_container(container=container)
-            # if same key is used multiple times, determine
-            # which instance based on the Container
-            object_field = self._check_object_field(file=file,
-                                                    container=container,
-                                                    relative_path=relative_path,
-                                                    field=field)
-            for row_idx in self.object_keys.which(objects_idx=object_field.idx):
-                key_idx = self.object_keys['keys_idx', row_idx]
-                if key_idx in key_idx_matches:
-                    return self.keys.row[key_idx]
-            msg = "No key found with that container."
-            raise ValueError(msg)
+        if file is None:
+            file = self._get_file_from_container(container=container)
+        # if key is provided then add_ref proceeds as normal
+        if key is not None:
+            data = [key]
         else:
-            if len(key_idx_matches) == 0:
-                # the key has never been used before
-                raise ValueError("key '%s' does not exist" % key_name)
-            elif len(key_idx_matches) > 1:
-                msg = "There are more than one key with that name. Please search with additional information."
-                raise ValueError(msg)
+            # if the key is not provided, proceed to "bulk add"
+            if attribute is None:
+                data_object = container
             else:
-                return self.keys.row[key_idx_matches[0]]
-
-    @docval({'name': 'entity_id', 'type': str, 'doc': 'The ID for the identifier at the resource.'})
-    def get_entity(self, **kwargs):
-        entity_id = kwargs['entity_id']
-        entity = self.entities.which(entity_id=entity_id)
-        if len(entity)>0:
-            return self.entities.row[entity[0]]
-        else:
-            return None
+                data_object = getattr(container, attribute)
+            if isinstance(data_object, (Data, DataIO)):
+                data = data_object.data
+            elif isinstance(data_object, (list, tuple, np.ndarray)):
+                data = data_object
+        missing_terms = []
+        for term in data:
+            # check the data according to the permissible_values
+            try:
+                term_info = termset[term]
+            except ValueError:
+                missing_terms.append(term)
+                continue
+            entity_id = term_info[0]
+            entity_uri = term_info[2]
+            self.add_ref(file=file,
+                         container=container,
+                         attribute=attribute,
+                         key=term,
+                         field=field,
+                         entity_id=entity_id,
+                         entity_uri=entity_uri)
+        if len(missing_terms)>0:
+            return {"missing_terms": missing_terms}
```

Comment:
I forget - why do we return a dict here with a single key-value pair instead of just the list of missing terms?

Reply:
My personal preference in that it gives information that it is indeed missing terms